feat(tasks): Task-tracking tool

orospakr opened the pull request:
This is an implementation of a task tracking tool in Crush similar to that found in other coding agents on the market.

---

<img width="1920" height="995" alt="Pasted image 20250812084930" src="https://github.com/user-attachments/assets/79213519-96d8-4eb7-99f5-7c09ac2fc583" />

Task tracking tools allow the model to plan its next steps, remain focused, and avoid forgetting tasks, just like a human software engineer would do with their notes. It also has the benefit of giving the user the opportunity to better track the agent's progress through each step of a job.

- Addition of three new LLM tools: `todo_write`, `todo_list`, and `todo_delete`
- TODO items tracked in the project-level (pwd) sqlite DB found in `.crush`
- Adds presentation of the TODOs to the sidebar
- Modifies the system prompts (further meta-prompting messaging may be needed for some of them?) for the various model vendors to encourage them to use the tools to plan ahead and track progress.
- An additional change: added DB testing helper routine for creating an in-memory sqlite DB with the Goose migrations applied, to enable tests for code that interact with the DB

Usage is just to use Crush as normal, and now the agent will plan out its work and use task tracking automatically, but for testing you can ask the model to directly add and remove TODOs.

naresh-tech-backend commented:
I've fixed the issue and also there are couple of other parsing issue which i also fixed let me know if i can create new pull request

orospakr commented:
> I've fixed the issue and also there are couple of other parsing issue which i also fixed let me know if i can create new pull request

Hey @naresh-tech-backend  issue did you identify? (btw the CI failures should be resolved now)

kujtimiihoxha commented:
@orospakr we have an idea of how the design for this should be but it looks like this could be a good base for the feature, will review this and maybe work ontop of this to match the design we want.

caarlos0 commented:
that sounds like a good idea to me!

i tested it, and got some errors though:

<img width="2540" height="860" alt="CleanShot 2025-09-19 at 14 48 19@2x" src="https://github.com/user-attachments/assets/6a0bd063-4d81-4b8f-9847-d8b936e48927" />

caarlos0 commented:
See also: https://github.com/charmbracelet/crush/discussions/494
